core/lib/soci/core/session.cpp

```diff
@@ -80,12 +80,20 @@ session::session(std::string const & connectString)
 session::session(connection_pool & pool)
     : logStream_(NULL), isFromPool_(true), pool_(&pool)
 {
-    poolPosition_ = pool.lease();
-    session & pooledSession = pool.at(poolPosition_);
-
-    once.set_session(&pooledSession);
-    prepare.set_session(&pooledSession);
-    backEnd_ = pooledSession.get_backend();
+    while (true) {
+        poolPosition_ = pool.lease();
+        session & pooledSession = pool.at(poolPosition_);
+
+        if (pooledSession.is_alive()) {
+            once.set_session(&pooledSession);
+            prepare.set_session(&pooledSession);
+            backEnd_ = pooledSession.get_backend();
+            break;
+        } else {
+            // invalidate the session in the pool, and try again
+            pool.give_back(poolPosition_);
+        }
+    }

```

Comment:
I suggest to create a new configuration value such as `MAX_RETRY_SESSION_CONNECTION` or something similar to avoid an infinity loop. 
Even if you dislike exception the case where we all the tries failed can be considered as exceptional.

Reply:
I agree.

Reply:
Since the change is `soci` related maybe the configuration value cannot be applied in that scope